docs/source/threats.rst

```diff
@@ -0,0 +1,192 @@
+   corruption or exfiltration of sensitive data.
+
+
+Prevention & design considerations
+----------------------------------
+
+Same-origin policy protections
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+All arXiv applications and services should prevent click-jacking attacks by
+setting the ``Content-Security-Policy`` and ``X-Frame-Options`` headers on all
+responses. For example:
+
+.. code-block:: python
+
+   response.headers["Content-Security-Policy"] = "frame-ancestors 'none'"
+   response.headers["X-Frame-Options"] = "SAMEORIGIN"
+
+This prevents arXiv pages from being displayed on other sites (e.g. via an
+iframe).
+
+Cross-Site Request Forgery protection
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+The risk of replay attacks or other attacks that involve generating fraudulent
+requests can be mitigated through CSRF protection. This usually involves
+issuing a token to the client that must be included in a subsequent request
+(e.g. when POSTing form data). The :mod:`wtforms` package provides `tools for
+CSRF protection
+<http://wtforms.readthedocs.io/en/latest/csrf.html#module-wtforms.csrf>`_,
+including an API for implementing a CSRF protection mechanism.
+
+.. todo::
+
+   Document CSRF implementation in :mod:`arxiv.base`.
+
+
+Input sanitization
+^^^^^^^^^^^^^^^^^^
+We use :mod:`wtforms` for all form-handling in arXiv-NG applications. By
+default, WTForms uses `MarkupSafe <https://github.com/pallets/markupsafe>`_ to
+escape unsafe HTML in all form input. This means that form data accessed in
+templates and other locations should be safe for display without risk of
+reflection XSS attacks.
+
+We use :mod:`sqlalchemy` to interact with the legacy database and other
+relational databases. We can minimize the risk of SQL injection attacks by
+never writing raw SQL statements, and instead using the SQLAlchemy ORM/DSL
+to build queries.
+
+Cookie security
+^^^^^^^^^^^^^^^
+Session keys and permanent login tokens that are stored in cookies should be
+set with the following attributes:
+
+- ``HttpOnly``: prevents access to the cookie by client-side scripts.
+- ``Secure``: the cookie should only be transmitted via HTTPS.
+- ``SameSite``: prevents sending the cookie with cross-site requests.
+
+Enforcement of these cookie policies is up to the user's browser.
+
+These attributes can be set when using the :meth:`flask.Response.set_cookie`
+API. For example:
+
+.. code-block:: python
+
+   response = make_response(...)
+   response.set_cookie('foocookie', 'secret value', secure=True,
+                       httponly=True, samesite=True)
+
+In addition, both legacy and NG session cookies are signed with a secret hash.
+
+In the former case, the cookie is verified by generating a digest from its
+contents along with a salt and comparing it to the digest transmitted in the
+cookie itself.
+
+In the latter case, the cookie is an encrypted JWT that must be decrypted
+successfully using a secret key. In addition, a **pseudo-random nonce** created
```

Comment:
For my edification - using the secret key means that the cookie's value cannot be determined by looking at source in our public repos, among other benefits? Or is that not a concern because of the way cookie values are created in the first place?

Reply:
Yeah, even if you know the encryption algorithm (e.g. sha256), the only way to figure out what is in the token is if you know the secret key that was used to generate it.